sw/host/opentitantool/src/command/console.rs

```diff
@@ -0,0 +1,246 @@
+                None
+            };
+            console.interact(transport, &mut stdin, &mut stdout)?;
+        }
+        if !self.quiet {
+            println!("\n\nExiting interactive console.");
+        }
+        Ok(None)
+    }
+}
+
+#[derive(Default)]
+struct InnerConsole {
+    logfile: Option<File>,
+    deadline: Option<Instant>,
+    exit_success: Option<Regex>,
+    exit_failure: Option<Regex>,
+    buffer: String,
+}
+
+enum ExitStatus {
+    None,
+    ExitSuccess,
+    ExitFailure,
+}
+
+impl InnerConsole {
+    const CTRL_C: u8 = 3;
+    const BUFFER_LEN: usize = 1024;
+
+    // Runs an interactive console until CTRL_C is received.
+    fn interact<R>(
+        &mut self,
+        transport: &mut dyn Transport,
+        stdin: &mut R,
+        stdout: &mut impl Write,
+    ) -> Result<()>
+    where
+        R: Read + AsRawFd,
+    {
+        let mut uart = transport.uart()?;
+        let mut buf = [0u8; 256];
+
+        loop {
+            if let Some(deadline) = self.deadline {
+                if Instant::now() > deadline {
+                    // If we have an exit success condition, then a timeout
+                    // should be an error.
+                    if self.exit_success.is_some() {
+                        return Err(anyhow!("Console timeout exceeded"));
+                    } else {
+                        break;
+                    }
+                }
+            }
+
+            // This loop _should_ really use `poll` to learn when the
+            // console or uart file descriptors become ready, but some UART
+            // backends will bury their implementation in libusb and make
+            // discovering the file descriptor difficult or impossible.
+            //
+            // As a pragmatic implementation detail, we wait for the UART
+            // for a short period of time and then service the console.
+
+            // Check for input on the uart.
+            match self.uart_read(&mut *uart, Duration::from_millis(10), stdout)? {
+                ExitStatus::None => {}
+                ExitStatus::ExitSuccess => {
+                    break;
+                }
+                ExitStatus::ExitFailure => {
+                    return Err(anyhow!("Matched exit_failure expression"));
+                }
+            };
+
+            // Wait for input from the user.
+            match file::wait_read_timeout(&*stdin, Duration::from_millis(0)) {
+                Ok(_) => {
+                    let len = stdin.read(&mut buf)?;
+                    if len == 1 && buf[0] == InnerConsole::CTRL_C {
+                        break;
+                    }
+                    uart.write(&buf[..len])?;
+                }
+                Err(_) => {
+                    // Can only be timeout.  Do nothing.
+                }
+            };
+        }
+        Ok(())
+    }
+
+    // Maintain a buffer for the exit regexes to match against.
+    fn append_buffer(&mut self, data: &[u8]) {
```

Comment:
How about this, instead (assuming `buffer` is a `Vec<u8>`, which I feel is somewhat more sensible):
```rust
if self.buffer.len() + data.len() > BUFFER_LEN {
  let new_len = BUFFER_LEN - data.len();
  self.buffer.copy_within(buffer.len() - new_len.., 0);
  self.buffer.trunc(new_len);
}
```
and then just append as before.

The fussy bit is that str doesn't have copy_within for obvious reasons (to avoid slicing through surrogates, among other things), but `Regex` needs a string. I think the thing to do there is use `from_utf8`? Maybe this isn't worth it, shrug.